zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub struct SigHasher<'a> {
+    pub(crate) trans: &'a Transaction,
+    pub(crate) hash_type: u32,
+    pub(crate) network: Network,
+    pub(crate) height: BlockHeight,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unimplemented!(),
+            BeforeOverwinter => unimplemented!(),
+            Overwinter | Sapling => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Blossom => unimplemented!(),
+            Heartwood => unimplemented!(),
+            Canopy => unimplemented!(),
+        }
+
+        hash.finalize()
+    }
+
+    fn network_upgrade(&self) -> NetworkUpgrade {
+        NetworkUpgrade::current(self.network, self.height)
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade().branch_id().unwrap()
+    }
+
+    /// Sighash implementation for the overwinter and sapling consensus branches
+    fn hash_sighash_zip143<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.hash_header(&mut writer)?;
+        self.hash_groupid(&mut writer)?;
+        self.hash_prevouts(&mut writer)?;
+        self.hash_sequence(&mut writer)?;
+        self.hash_outputs(&mut writer)?;
+        self.hash_joinsplits(&mut writer)?;
+        self.hash_lock_time(&mut writer)?;
+        self.hash_expiry_height(&mut writer)?;
+        self.hash_hash_type(&mut writer)?;
+
+        // if let Some((n, script_code, amount)) = transparent_input {
+        //     let mut data = vec![];
+        //     tx.vin[n].prevout.write(&mut data).unwrap();
+        //     script_code.write(&mut data).unwrap();
+        //     data.extend_from_slice(&amount.to_i64_le_bytes());
+        //     (&mut data)
+        //         .write_u32::<LittleEndian>(tx.vin[n].sequence)
+        //         .unwrap();
+        //     h.update(&data);
+        // }
+
+        Ok(())
+    }
+
+    pub(super) fn personal(&self) -> [u8; 16] {
+        let mut personal = [0; 16];
+        (&mut personal[..12]).copy_from_slice(ZCASH_SIGHASH_PERSONALIZATION_PREFIX);
+        (&mut personal[12..])
+            .write_u32::<LittleEndian>(self.consensus_branch_id().into())
+            .unwrap();
+        personal
+    }
+
+    fn hash_header<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(self.trans.header())
+    }
+
+    fn hash_groupid<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer
+            .write_u32::<LittleEndian>(self.trans.group_id().expect("fOverwintered is always set"))
```

Comment:
Question: What stops this code being run on a v1 or v2 transaction?
```suggestion
            .write_u32::<LittleEndian>(self.trans.group_id().expect("Transaction is V3 or later"))
```

Reply:
possibly nothing, I'm still figuring things out.